cjwkernel/pandas/types.py

```diff
@@ -411,19 +411,41 @@ class I18nMessage:
     args: Dict[str, Union[int, float, str]] = field(default_factory=dict)
     """Arguments (empty if message does not need any -- which is common)."""
 
-    source: Dict[str, str] = field(default_factory=dict)
+    source: Optional[Dict[str, str]] = None
     """An indication of where the message is coming from.
-        - An empty dict means it's coming from workbench itself
-        - A dict with key `"module_id"` means it's coming from a module
-        - A dict with key `"library"` indicates it's coming from some of our supported libraries (e.g. `"cjwmodule"`)
+        - `None` means it's coming from workbench itself
+        - A dict with key `"module"` means it's coming from a module
+          for example `{"module": "mymodule"}` means, "search the mymodule module's catalog"
+        - A dict with key `"library"` indicates it's coming from some of our supported libraries;
+          for example `{"library": "cjwmodule"}` means, "search the cjwmodule library's catalog"
     """
 
     @classmethod
     def from_arrow(cls, value: atypes.I18nMessage) -> I18nMessage:
-        return cls(value.id, value.args, value.source)
+        if not value.source:
+            return cls(value.id, value.args)
+        if isinstance(value.source, atypes.I18nMessageSource.Module):
+            return cls(value.id, value.args, {"module": value.source.module_id})
+        if isinstance(value.source, atypes.I18nMessageSource.Library):
+            return cls(value.id, value.args, {"library": value.source.library})
+        raise ValueError("Unknown I18nMessage source")
```

Comment:
Don't tempt callers to recover from undefined behavior.

`ValueError` suggests to me that somewhere, somehow, a _user_ entered the wrong data. And it suggests that a caller may want to catch the error.

But if this exception is raised, _cjworkbench itself_ is buggy. So if there _must_ be an exception, make it one that nobody would _dare_ catch -- for instance, `assert False` (AssertionError) or `raise NotImplementedError` (a RuntimeError).

In this specific function, there's a better approach: rearrange the if-statements so the `return cls(value.id, value.args)` comes in an `else:` clause.

Try to picture an integration-test suite that leads to 100% code coverage. It will never hit this `raise` line. So the line shouldn't exist.